### ISSUE TITLE
 Proposing Huginn.io - Homepage and Scenario sharing site for Huginn 

### ISSUE THREAD

**Issue by dsander** [cantino/huginn#1774]
@Enfop and I have been working on working on a Huginn Community/Homepage/Scenario sharing page in the last few weeks.

A demo deployment is running on Heroku: [http://huginnio.herokuapp.com/](http://huginnio.herokuapp.com/). 
Be aware: we might need to move the installation or wipe the database while the development continues.

It initially started as an implementation of the idea behind [Communnin](https://github.com/omniscopeio/communnin) but then shifted into a more general "Homepage" / "one stop shop" to learn Huginn.

The site currently has four main sections:

#### Documenation

The idea is to have a condensed version of the README, Wiki articles and links to the documentation of the Huginn repository.

Possible sections:

* Introduction to Huginn
  * Sample use cases
  * Comparison with other services/OSS software
* Concepts
  * Agents
    * Options
    * Liquid
    * JSONPath
  * Events
  * Scenarios
  * Services
  * Credentials
  * Agent Gems
* Guides, possibly user submitted?

#### Agent Database

The Agent Database contains a searchable list of all Huginn Agents with their features and descriptions. It allows users to get an overview of what is possible without having to install Huginn first.

Possible extensions:

* Add Agent "categories" like "Utility" (Trigger, DeDuplication, Formatting, ...), "Service Consumer" (Basecamp, Imap, Twitter*, ...), etc
* Filter Agents by capabilities (creates Events, receives Events, etc), Service, File Pointer

#### Scenarios

The "heart" of Huginn.io, users can upload their Scenarios and discover new ones.

Uploading Scenarios works like importing them into Huginn, either via JSON files or by specifying a public Scneario URL. Scenarios are searchable and should be filterable by Agents/Services/features in the future.

Possible extensions:

* Versioned Scenarios
* Comments/Discussions
* Rating
* Integration with Huginn to auto update Scenarios both from and to Huginn.io

#### Agent Gems

Auto-updated, searchable list of all Agent Gems on Github. Makes the discovery of (new) Agent Gems easier then browsing through GitHub manually.

The code is currently [hosted](https://git.dsander.de/dxm/communin) on a private GitLab server, but will eventually move to GitHub.

There is still a lot to do, but given a few more weeks and maybe some help in form of PRs we should be able to get an initial version done pretty soon.

We hope the community likes our approach and accepts the idea as the "official" Huginn Homepage.

**Comment by cantino** [comment]
I am a *HUGE* fan of this. So, so great!

Maybe the Agent Database should be inside Huginn itself too? Could we share the code?

We should definitely link from Huginn to this once it's deployed, and allow Scenario subscriptions.

**Comment by signallhill** [comment]
Its great. Great work.

**Comment by bobbysteel** [comment]
Great idea!

**Comment by vivagonna** [comment]
Love this idea. Great work!

**Comment by dsander** [comment]
> Maybe the Agent Database should be inside Huginn itself too? Could we share the code?

It's not much code involved in the Database, mostly the template and the search function (which could be tricky to port to mysql especially when we improve it from the current simple form), otherwise it should be simple.

**Comment by JamiePhonic** [comment]
Looks like i ran into an issue importing a scenario from a public link, but importing the file worked fine.
so far so good!!

**Comment by dsander** [comment]
@JamiePhonic Thanks for testing the scenario upload, did you get the "oops something went wrong" page when trying to import a scenario via a public link? If that was the case could you try it again? We now have exception tracking in place and it could help us to fix the bug.

**Comment by JamiePhonic** [comment]
@dsander I did origonally, but i imported another scenario just a few minutes ago using a public link without issues, so it looks like you fixed whatever the issue was.

Quick question: Is there an expiry time on modifying submitted scenarios? it appears i cant edit (well... delete) the "UK TV Air Dates" scenario i uploaded last week (i signed in using github if that helps)

**Comment by dsander** [comment]
@JamiePhonic There isn't an expiry time, but a caching issue. I'll fix that later, thanks for noticing.

**Comment by dsander** [comment]
@JamiePhonic You should now be able to delete your scenario again.

**Comment by JamiePhonic** [comment]
Yep, all fixed.

I take it there will be a way of updating a scenario instead of deleting and re-uploading it in the future?
Perhaps a way of automatically pulling updates across if people used a public link to upload a scenario instead of a file? (could be a selectable option on the upload screen?)

(Just some random idea's, dont mind me :))

**Comment by dsander** [comment]
It's not clear in the UI at the moment, but you can reupload the updated scenario JSON to upload it. The reason of not having an edit form is to avoid conflicts. We think the scenario JSON (either uploaded or synced via an URL) should be the only source to change the scenario.
Once everything is implemented updating a scenario should work automatically from the publishers Huginn instance via Huginn.io to the subscribers instance (which fetches changes periodically).

**Comment by cantino** [comment]
I've been doing some testing and found a few issues.

* Before signing up, I clicked the "Sign in with GitHub" link and ended up at an empty page.
* After making an account, I was dropped on `/scenario_imports/new`. After navigating away, I couldn't find a link back until a bit later when I noticed the "New Scenario" sidebar link. An icon would be useful to call this out.
* It'd be nice if downloaded Scenario JSON files were generated with `pretty_generate` for readability.
* I assume "Copy Scenario URL" will eventually copy instead of download.

All minor.  Great work so far!

Are you going to move this to GitHub soon? I'm happy to start sending PRs ;)

**Comment by dsander** [comment]
Sorry for the late response I lost track about Huginn.io for a bit. Actually I moved both the rails application and the agent description/gem updater to GitHub a few weeks ago:
https://github.com/dsander/huginn.io
https://github.com/dsander/huginnio_agent_updater

The updater is running every day on Travis to keep the site up to date.

PRs are of course welcome, I will try to convert my TODO list into issues and incorporate your feedback in the next days.

**Comment by Phyks** [comment]
Wow, this looks really promising! Having the "agents database" page inside Huginn could be really helpful I think, to be able to have an overview of all available (installed) agents.

**Comment by bobbysteel** [comment]
Would be neat to either allow commenting on each scenario page or at least allow web-editable notes from the author. You can easily envision a time when there are multiple "competing" scenarios to do the same thing, so will be useful to have a way to discuss and or point to more recent versions (i.e. if someone forks and revs to v2 etc).

**Comment by dsander** [comment]
I agree comments would be a nice feature, right now I am struggling to get the first usable version ready. Besides some minor changes it's mainly the missing docs that I would like to get done.

**Comment by bobbysteel** [comment]
Fair enough :)

**Comment by waldens** [comment]
Many thanks to @dsander . Just upload a few test scenarios, most of the features needed are already available.

I feel the next needed best is the editing function. The scenario I uploaded have a bit wrong, but can not be modified. Also, when I use the Public Scenario URL to create a new scenario, the error is reported: `We're sorry, but something went wrong.`

Again, excellent work!

**Comment by dsander** [comment]
@waldens Thanks for testing it. The import issue might be a bug in Huginn when dealing with Scenarios and UTF-8 characters. Can you open a new issue for that and include the backtrace from you `production.log`?

I explained the reasoning why we did not add a edit function yet in a [comment above](https://github.com/cantino/huginn/issues/1774#issuecomment-264291179). We might add it later, but handling both automatic updates from Huginn and manual edits will be tricky.

**Comment by popovn** [comment]
idea is dead?
